### ISSUE TITLE
Component::setStateVariableValues() enforces constraints with every setValue()

### ISSUE THREAD

**Issue by tkuchida** [opensim-org/opensim-core#1884]
From discussion with @chrisdembia.

When setting the values of Coordinates, we want to enforce constraints only after all Coordinates have been set. Enforcing constraints with each `setValue()` is not just inefficient: it will also be mangling the newly set Coordinate values along the way, resulting in a final pose that could be far from what's expected. [This code](https://github.com/opensim-org/opensim-core/blob/master/OpenSim/Tools/CMCTool.cpp#L652) does it correctly. `Component::setStateVariableValues()` [seems to be enforcing constraints with every call to `setValue()`](https://github.com/opensim-org/opensim-core/blob/master/OpenSim/Common/Component.cpp#L944). I think Component.cpp#L944 needs to be
```cpp
_allStateVariables[i]->setValue(state, values[i], false);
```
when setting the value of a Coordinate, with a call to project/realize after L945.

**Comment by chrisdembia** [comment]
I would really like to discuss this in the next dev meeting.

**Comment by chrisdembia** [comment]
Here are some options for solving this issue (discussed with @tkuchida):

1. Have `Coordinate::CoordinateStateVariable::setValue()` pass `false` for the `enforceConstraints` parameter of `Coordinate::setValue()`. We would edit the documentation of `Component::setStateVariableValues()` to say that users must call `model.assemble()`, etc. on their own to enforce the constraints.
2. Have `Component::setStateVariableValues()` take a `enforceConstraints` parameter, and enforce the constraints only after setting all state variable values. This could be achieved by giving `Component` a virtual function that `Model` would implement by calling `assemble()`. I don't really like this idea. How is a component supposed to know how to enforce constraints?
3. Permit users to have a way to set state variable values without using `Component::setStateVariableValues()`. That is, help users use `State::updY()` directly. I see 2 options:

    1. provide a method that gives the names of the state variables in the order of `State::getY()`.
    2. Implement `Component::StateVariable::getSystemYIndex()` and expose these global indices to users, perhaps via `Component::getSystemYIndices()` (similar to `Component::getStateVariableNames()`).

**Comment by tkuchida** [comment]
Would it make more sense for `Coordinate::setValue()` to have `enforceConstraints = false` by default? I recognize that this change might cause some commotion for backwards compatibility (though we also reversed the sense of the `isDisabled` property). It seems that it would be more likely to accidentally enforce constraints when setting each Coordinate in a model (e.g., as scripting becomes more popular) compared to accidentally forgetting to assemble after setting all Coordinates. The former mistake would go unnoticed if the Coordinate values aren't checked; I think the latter mistake would throw (and, hopefully, provide a helpful message to the user).

We might also consider removing the `enforceConstraints` flag altogether. The user needs to be aware of whether constraints are being enforced regardless; explicitly calling something like `Model::enforceConstraints()` might be a better way to convey this information than setting a flag on the last call to `setValue()`.

**Comment by chrisdembia** [comment]
Perhaps we could discuss that in the dev meeting tomorrow.

**Comment by tkuchida** [comment]
> Perhaps we could discuss that in the dev meeting tomorrow.

I won't be at the meeting. Here are two ideas I didn't include in my comment above:

1. Having `setValue()` enforce constraints seems similar to having `setFrameName()` set the Frame itself rather than just the name (as suggested by the name of the method; see #1889). Following this logic, `setValue()` should just set the value—or, at least, just set the value by default (i.e., `enforceConstraints = false` by default).

2. I think the 4.0 design philosophy says that we should iterate through all Coordinates like this (rather than use Model's CoordinateSet):
```cpp
for (const auto& coord : model.getComponentList<Coordinate>()) {
    //do stuff
}
```
However, if `do stuff` includes setting values, I don't know how to determine when you've reached the last item in the ComponentList without adding a counter (and counting the number of Coordinates beforehand). I know there's a [`ComponentList<T>::end()` method](http://myosin.sourceforge.net/1902/classOpenSim_1_1ComponentList.html#ab2944d26aaa87050660b5fb11565c88d), but the doxygen says that it "points past the end of the list, *not* to the last item in the list" so, to determine whether you're setting the value of the last Coordinate, it seems like you might need to do something like
```cpp
const auto& coordList = model.getComponentList<Coordinate>();
for (const auto& coord : coordList) {
    const bool enforceConstraints = (coord.next().next() == coordList.end());
    coord.setValue(state, value, enforceConstraints);
}
```
which doesn't seem so clean. A counterargument is that the `value` being set would need to come from something, and presumably there would be as many elements in that "something" as there are Coordinates (thereby providing the signal of when `enforceConstraints` should be true).

**Comment by chrisdembia** [comment]
We decided not to change the default value of `enforceConstraints` because violating the constraints does not actually throw an error message and could lead to silent errors. 

Also, @aseth1 will address this issue by caching the state indices instead of the StateVariable objects. We also need to fix how we detect if the system is changed; we can add a virtual method on Component that Model invokes if the System has changed (perhaps `clearCachesFromSystem()` or `respondToSystemChanged()`). Actually, perhaps Model's `extendAddToSystemAferSubcomponents()` simply needs to clear and rebuild the state index cache.

@aseth1, I have a branch where I compute the System Y indices; see https://github.com/opensim-org/opensim-core/compare/implicit_operator#diff-a8c6cc5aa99c4c61fcc4d4b0a45a4813R2397

**Comment by tkuchida** [comment]
> violating the constraints does not actually throw an error message and could lead to silent errors

That's interesting. There should be a user-facing warning in the doxygen for `Coordinate::setValue()`.

**Comment by chrisdembia** [comment]
I had thought that realizePosition() would cause the constraints to be enforced, but @sherm1 was saying that you need to explicitly enforce the constraints (with the assembly solver, with `projectQ()`, etc.) and that realizing to any stage simply uses the states as they are given.


> There should be a user-facing warning in the doxygen for Coordinate::setValue().

Yes, I agree. The doxygen should also suggest calling `model.assemble()` and passing `enforceConstraints = false` if setting coordinate values in a loop.

**Comment by aseth1** [comment]
closed by #1911
